Haptics/app/src/main/java/com/example/android/haptics/samples/ui/views/HomeViewModel.kt

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.haptics.samples.ui.views
+
+import android.app.Application
+import android.os.Build
+import android.os.VibrationEffect
+import android.os.Vibrator
+import android.view.HapticFeedbackConstants
+import android.view.View
+import androidx.compose.foundation.ScrollState
+import androidx.compose.material.ScaffoldState
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.example.android.haptics.samples.R
+import kotlinx.coroutines.launch
+import java.lang.RuntimeException
+
+/**
+ * The state object the home screen needs to render the haptic feedback categories and buttons.
+ */
+data class HomeViewModelState(val hapticFeedbackCategories: List<HapticFeedbackCategory> = emptyList())
+
+/**
+ * Represents a category of haptic feedback APIs on the home screen.
+ */
+data class HapticFeedbackCategory(
+    val label: String = "", // The label displayed before the rows of buttons.
+    val categoryType: HapticFeedbackCategoryType,
+    val buttons: List<HapticFeedbackButton> = emptyList()
+)
+
+/**
+ * A single button on the home page, and the corresponding information we need to react to a click.
+ */
+data class HapticFeedbackButton(
+    val label: String = "",
+    // On some devices, effect won't work.
+    val worksOnUserDevice: Boolean,
+    // The corresponding constant for the various vibration effects available to developers.
+    val vibrationEffectConstant: Int = -1
+)
+
+/**
+ * The category type, necessary since each haptic feedback effect calls the APIs differently.
+ */
+enum class HapticFeedbackCategoryType {
+    PREDEFINED_EFFECTS, HAPTIC_CONSTANTS, PREDEFINED_PRIMITIVES
```

Comment:
Maybe rename a few of them to be closer to the public APIs:

COMPOSITION_PRIMITIVES -> VibrationEffect.Composition.addPrimitive()
PREDEFINED_EFFECTS -> VibrationEffect.createPredefined()
HAPTIC_FEEDBACK_CONSTANTS -> HapticFeedbackConstant